src/Tools/Pagination/Paginator.php

```diff
@@ -171,13 +206,25 @@ private function cloneQuery(Query $query): Query
     /**
      * Determines whether to use an output walker for the query.
      */
-    private function useOutputWalker(Query $query): bool
+    private function useOutputWalker(Query $query, bool $forCountQuery = false): bool
     {
-        if ($this->useOutputWalkers === null) {
-            return (bool) $query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) === false;
+        if ($this->useOutputWalkers !== null) {
+            return $this->useOutputWalkers;
         }
 
-        return $this->useOutputWalkers;
+        // When a custom output walker already present, then do not use the Paginator's.
+        if ($query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) !== false) {
+            return false;
+        }
+
+        // When not joining onto *ToMany relations, then do not use the more complex CountOutputWalker.
+        return ! (
+            $forCountQuery
```

Comment:
if we have a separate detection for count query and result query, I would rather extract the check for `$forCountQuery` in a separate query (as maybe we will also have smarter detection for the case of the result query in the future)

Reply:
You said:
> extract the check for $forCountQuery in a separate query

Did you mean to say "extract (...) in a separate _function_"? In other words, are you proposing splitting the `useOutputWalker()` into `useResultOutputWalker()` and `useCountOutputWalker()`?

Reply:
IIRC, I meant `a separate condition` (no need for a separate function which would force to duplicate the previous logic. We could even factor out the auto-detection check:

This would look like that:

```php
// ...

if (!$this->queryStyleAutoDetectionEnabled) {
    return true; // Old simple guessing
}

if ($forCountQuery) {
    return /* condition for the support of CountWalker */;
}

return /* condition for the support of LimitSubqueryWalker */;
```